Add apply_activation_checkpointing helper function

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #78854
* #78752
* __->__ #78704


Adds `apply_activation_checkpointing` helper function, which will allow users to easily checkpoint multiple submodules within a module depending on some condition. 

The function takes in `model`, `checkpoint_wrap_fn` which indicates the checkpoint_wrapper config, and a lambda indicating whether a module should be wrapped or not. It does this wrapper and updates the model in place, i.e.

`Sequential(Linear, Linear)` would become `Sequential(wrapped(Linear), wrapped(Linear))` in place. 

This is useful for several uses cases we've been seeing that have been trying to apply activation checkpointing, where we've had to hack around the module hierarchy via setattr/getattr which can easily lead to issues if not done properly.

rohan-varma committed:
Checkpoint util

[ghstack-poisoned]

rohan-varma committed:
Update on "Add apply_activation_checkpointing helper function"



Adds `apply_activation_checkpointing` helper function, which will allow users to easily checkpoint multiple submodules within a module depending on some condition. 

The function takes in `model`, `checkpoint_wrap_fn` which indicates the checkpoint_wrapper config, and a lambda indicating whether a module should be wrapped or not. It does this wrapper and updates the model in place, i.e.

`Sequential(Linear, Linear)` would become `Sequential(wrapped(Linear), wrapped(Linear))` in place. 

This is useful for several uses cases we've been seeing that have been trying to apply activation checkpointing, where we've had to hack around the module hierarchy via setattr/getattr which can easily lead to issues if not done properly.

[ghstack-poisoned]

rohan-varma committed:
Update on "Add apply_activation_checkpointing helper function"



Adds `apply_activation_checkpointing` helper function, which will allow users to easily checkpoint multiple submodules within a module depending on some condition. 

The function takes in `model`, `checkpoint_wrap_fn` which indicates the checkpoint_wrapper config, and a lambda indicating whether a module should be wrapped or not. It does this wrapper and updates the model in place, i.e.

`Sequential(Linear, Linear)` would become `Sequential(wrapped(Linear), wrapped(Linear))` in place. 

This is useful for several uses cases we've been seeing that have been trying to apply activation checkpointing, where we've had to hack around the module hierarchy via setattr/getattr which can easily lead to issues if not done properly.

[ghstack-poisoned]

awgu commented on torch/distributed/algorithms/_checkpoint/checkpoint_wrapper.py:
```diff
@@ -123,3 +123,54 @@ def checkpoint_wrapper(
+    Returns: None (`model` is modified inplace)
+    """
+    layers = []
+    for name, layer in model.named_modules():
+        if check_fn(model, layer):
+            toks = name.strip().split(".")
+            module = layer
+            layer = model
```
Some interesting variable aliasing going on here 😆 

```suggestion
    for name, module in model.named_modules():
        if check_fn(model, module):
            toks = name.strip().split(".")
            layer = model
```

awgu replied:
Overall, this method is a bit hard to follow for me because the variables keep changing what they mean.

It seems like the point is that we have to find `nn.Module` attributes of an `nn.Module` to make sure to set them as attributes of the new `nn.Module` returned by `checkpoint_wrapper_fn()`?

`checkpoint_wrapper_fn()` is only applied to the original `nn.Module`s of `model` (via `new_layer = checkpoint_wrapper_fn(module)` where `module = layer` and `layer` is from `model.named_modules()`).

Why do we only need to `setattr` for `nn.Module` attributes for the new `nn.Module` returned by `checkpoint_wrapper_fn()`? Why are they different from other non-`nn.Module` attributes?

rohan-varma replied:
> Why do we only need to setattr for nn.Module attributes for the new nn.Module returned by checkpoint_wrapper_fn()? Why are they different from other non-nn.Module attributes?

We only need to set these since these are new modules, i.e. a `CheckpointWrapper` module containing a submodule which is the wrapped module.

The rest of the modules we didn't wrap are already set? 

Agreed, naming here can definitely be better and function can be made clearer, let me update.

rohan-varma committed:
Update on "Add apply_activation_checkpointing helper function"



Adds `apply_activation_checkpointing` helper function, which will allow users to easily checkpoint multiple submodules within a module depending on some condition. 

The function takes in `model`, `checkpoint_wrap_fn` which indicates the checkpoint_wrapper config, and a lambda indicating whether a module should be wrapped or not. It does this wrapper and updates the model in place, i.e.

`Sequential(Linear, Linear)` would become `Sequential(wrapped(Linear), wrapped(Linear))` in place. 

This is useful for several uses cases we've been seeing that have been trying to apply activation checkpointing, where we've had to hack around the module hierarchy via setattr/getattr which can easily lead to issues if not done properly.

[ghstack-poisoned]

rohan-varma committed:
Update on "Add apply_activation_checkpointing helper function"



Adds `apply_activation_checkpointing` helper function, which will allow users to easily checkpoint multiple submodules within a module depending on some condition. 

The function takes in `model`, `checkpoint_wrap_fn` which indicates the checkpoint_wrapper config, and a lambda indicating whether a module should be wrapped or not. It does this wrapper and updates the model in place, i.e.

`Sequential(Linear, Linear)` would become `Sequential(wrapped(Linear), wrapped(Linear))` in place. 

This is useful for several uses cases we've been seeing that have been trying to apply activation checkpointing, where we've had to hack around the module hierarchy via setattr/getattr which can easily lead to issues if not done properly.

[ghstack-poisoned]

awgu commented on torch/distributed/algorithms/_checkpoint/checkpoint_wrapper.py:
```diff
@@ -123,3 +123,56 @@ def checkpoint_wrapper(
+            # This submodule should be wrapped with activation checkpointing
+            tokens = name.strip().split(".")
+            submodule_parent = model
+            # Get parent layer of module by iterating through the tokenized FQN.
```
nit: (IIUC)
```suggestion
            # Get parent layer of submodule by iterating through the tokenized FQN.
```

awgu reviewed (commented):
Thanks for clearing this up. I think it makes sense now.

rohan-varma committed:
Update on "Add apply_activation_checkpointing helper function"



Adds `apply_activation_checkpointing` helper function, which will allow users to easily checkpoint multiple submodules within a module depending on some condition. 

The function takes in `model`, `checkpoint_wrap_fn` which indicates the checkpoint_wrapper config, and a lambda indicating whether a module should be wrapped or not. It does this wrapper and updates the model in place, i.e.

`Sequential(Linear, Linear)` would become `Sequential(wrapped(Linear), wrapped(Linear))` in place. 

This is useful for several uses cases we've been seeing that have been trying to apply activation checkpointing, where we've had to hack around the module hierarchy via setattr/getattr which can easily lead to issues if not done properly.

[ghstack-poisoned]

zhaojuanmao reviewed (commented):
Wondering whether we can just use FSDP '_recursive_wrap()' API for this helper function as well? it also traverses the model, wrap submodules and change model in place. basically make the wrap_cls=checkpoint_wrapper?

zhaojuanmao commented on torch/distributed/algorithms/_checkpoint/checkpoint_wrapper.py:
```diff
@@ -123,3 +123,56 @@ def checkpoint_wrapper(
+            tokens = name.strip().split(".")
+            submodule_parent = model
+            # Get parent layer of module by iterating through the tokenized FQN.
+            for token in tokens[:-1]:
+                # isnumeric() check to handle case where layers are in a list,
+                # see https://discuss.pytorch.org/t/how-to-replace-a-layer-with-own-custom-variant/43586/12
+                if not token.isnumeric():
+                    submodule_parent = getattr(submodule_parent, token)
+                else:
+                    submodule_parent = submodule_parent[int(token)]
```
not sure the way to get submodule_parent is robust and general

rohan-varma committed:
Update on "Add apply_activation_checkpointing helper function"



Adds `apply_activation_checkpointing` helper function, which will allow users to easily checkpoint multiple submodules within a module depending on some condition. 

The function takes in `model`, `checkpoint_wrap_fn` which indicates the checkpoint_wrapper config, and a lambda indicating whether a module should be wrapped or not. It does this wrapper and updates the model in place, i.e.

`Sequential(Linear, Linear)` would become `Sequential(wrapped(Linear), wrapped(Linear))` in place. 

This is useful for several uses cases we've been seeing that have been trying to apply activation checkpointing, where we've had to hack around the module hierarchy via setattr/getattr which can easily lead to issues if not done properly.

[ghstack-poisoned]

rohan-varma commented:
> Wondering whether we can just use FSDP '_recursive_wrap()' API for this helper function as well? it also traverses the model, wrap submodules and change model in place. basically make the wrap_cls=checkpoint_wrapper?

It might require our own policy, as the user might want a more general lambda function? It can be quite similar to transformer_auto_wrap_policy.

rohan-varma committed:
Update on "Add apply_activation_checkpointing helper function"



Adds `apply_activation_checkpointing` helper function, which will allow users to easily checkpoint multiple submodules within a module depending on some condition. 

The function takes in `model`, `checkpoint_wrap_fn` which indicates the checkpoint_wrapper config, and a lambda indicating whether a module should be wrapped or not. It does this wrapper and updates the model in place, i.e.

`Sequential(Linear, Linear)` would become `Sequential(wrapped(Linear), wrapped(Linear))` in place. 

This is useful for several uses cases we've been seeing that have been trying to apply activation checkpointing, where we've had to hack around the module hierarchy via setattr/getattr which can easily lead to issues if not done properly.

[ghstack-poisoned]

zhaojuanmao commented on torch/distributed/fsdp/wrap.py:
```diff
@@ -29,6 +29,43 @@ def always_wrap_policy(*args, **kwargs) -> bool:
+       unwrapped_params (int):
+           The number of parameters yet to be wrapped in this module.
+
+       transformer_layer_cls (int):
```
nit: no 'transformer_layer_cls'?

zhaojuanmao reviewed (approved):
some minors, looks clean:)

rohan-varma committed:
Update on "Add apply_activation_checkpointing helper function"



Adds `apply_activation_checkpointing` helper function, which will allow users to easily checkpoint multiple submodules within a module depending on some condition. 

The function takes in `model`, `checkpoint_wrap_fn` which indicates the checkpoint_wrapper config, and a lambda indicating whether a module should be wrapped or not. It does this wrapper and updates the model in place, i.e.

`Sequential(Linear, Linear)` would become `Sequential(wrapped(Linear), wrapped(Linear))` in place. 

This is useful for several uses cases we've been seeing that have been trying to apply activation checkpointing, where we've had to hack around the module hierarchy via setattr/getattr which can easily lead to issues if not done properly.

[ghstack-poisoned]

rohan-varma committed:
Update on "Add apply_activation_checkpointing helper function"



Adds `apply_activation_checkpointing` helper function, which will allow users to easily checkpoint multiple submodules within a module depending on some condition. 

The function takes in `model`, `checkpoint_wrap_fn` which indicates the checkpoint_wrapper config, and a lambda indicating whether a module should be wrapped or not. It does this wrapper and updates the model in place, i.e.

`Sequential(Linear, Linear)` would become `Sequential(wrapped(Linear), wrapped(Linear))` in place. 

This is useful for several uses cases we've been seeing that have been trying to apply activation checkpointing, where we've had to hack around the module hierarchy via setattr/getattr which can easily lead to issues if not done properly.

[ghstack-poisoned]

rohan-varma commented:
@pytorchbot merge
